src/Filter/Widget/OptionMultiDynamicAggregate.php

```diff
@@ -151,23 +187,27 @@ private function getFilterQuery($terms)
     {
         list($parent, $child, $field) = explode('>', $this->getDocumentField());
         $boolQuery = new BoolQuery();
-
         foreach ($terms as $groupName => $values) {
             $innerBoolQuery = new BoolQuery();
 
             foreach ($values as $value) {
                 $nestedBoolQuery = new BoolQuery();
                 $nestedBoolQuery->add(new TermQuery($field, $value));
                 $nestedBoolQuery->add(new TermQuery($this->getNameField(), $groupName));
+
+                $inStockBoolQuery = new BoolQuery();
+
+                $inStockBoolQuery->add(new RangeQuery('variants.stock', ['gte' => 1]), BoolQuery::SHOULD);
```

Comment:
Shouldn't this be responsibility of `InStock` filter? Also in some cases it may make sense to allow searching in products that are no longer available for purchase.

Reply:
Blame how Elastic works - because those are nested documents, the contexts are different and it will return the document if it fields one with the defined option code regardless of it having stock or not.

Reply:
Oh, I see. Thanks for explaining the problem. The issue is when a product has e.g. red color variant in stock and blue color variant not in stock AND user in filter selects option[color] = blue. The `InStock` filter will return the product because the red variant is in stock and is unaware of the option filter requesting only blue variant.

The priceRange filter would have the same issue if `variants.price` (which may differ for each variant) would be used instead of `price` (currently computed as minimum of `variants.price`).

Reply:
Yep, exactly.

With price it's not the same because when we put the product into elastic, we take the minimum price from all the variants and put that as products price (price field on the ProductDocument) and that one is used for filtering. And because live index update is implemented, if the variant is sold out, the index is immediatly updated and so is the min price on the product, as is the stock. So it does not have that problem as options do.

Reply:
I was at this for a long few months, I have gone throught a lot of edge cases :)